### PR TITLE
Added ability to specify output names and types on any function using decorator

### PR DESCRIPTION
**Pull request by rdaly525**
This will enable automatically creating the coreir or magma type for a PE (or any peak "primitive")

### CODE REVIEW

File: peak/peak.py [hunk 1]
```diff
@@ -1,2 +1,45 @@
...
+
+def name_outputs(**outputs):
+    """Decorator meant to apply to any function to specify output types
+    The output types will be stored in fn.__peak_outputs__
+    The input types will be stored in fn.__peak_inputs__
+    Will verify that all the inputs have type annotations
+    Will also verify that the outputs of running fn will have the correct number of bits
+    """
+    def decorator(call_fn):
+        def call_wrapper(*args,**kwargs):
```
Comment (cdonovick) [inline peak/peak.py]: should decorate `call_wapper` with `@functools.wraps(call_fn)`
https://docs.python.org/3/library/functools.html#functools.wraps

File: peak/peak.py [hunk 2]
```diff
@@ -1,2 +1,45 @@
...
+            results = call_fn(*args,**kwargs)
+            single_output = not isinstance(results,tuple)
+            if single_output:
+                results = (results,)
+            for i, (oname,otype) in enumerate(outputs.items()):
+                assert results[i].num_bits == otype(0).num_bits, f"{results[i].num_bits} != {otype(0).num_bits}"
```
Comment (cdonovick) [inline peak/peak.py]: Should thrown a TypeError not raise an assertion.

Test should be `type(result[i]) is otype` after the merge of new BitVector.  Fine for now.

File: peak/peak.py [hunk 3]
```diff
@@ -1,2 +1,45 @@
...
+            return results
+
+        #Set all the outputs
+        call_wrapper.__peak_outputs__ = OrderedDict()
+        for oname,otype in outputs.items():
+            assert issubclass(otype,BitVector), f"{oname} is not a Bitvector class"
+            call_wrapper.__peak_outputs__[oname] = otype
+
+        #set all the inputs
+        call_wrapper.__peak_inputs__ = OrderedDict()
```
Comment (cdonovick) [inline peak/peak.py]: See above

File: peak/peak.py [hunk 4]
```diff
@@ -1,2 +1,45 @@
...
+                assert results[i].num_bits == otype(0).num_bits, f"{results[i].num_bits} != {otype(0).num_bits}"
+            if single_output:
+                results = results[0]
+            return results
+
+        #Set all the outputs
+        call_wrapper.__peak_outputs__ = OrderedDict()
+        for oname,otype in outputs.items():
+            assert issubclass(otype,BitVector), f"{oname} is not a Bitvector class"
+            call_wrapper.__peak_outputs__[oname] = otype
```
Comment (cdonovick) [inline peak/peak.py]: `__peak_outputs__` should be renamed.  Use of dunder names is reversed for magic methods and attributes by PEP 8.
https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles

File: peak/peak.py [hunk 5]
```diff
@@ -1,2 +1,45 @@
...
+
+        #set all the inputs
+        call_wrapper.__peak_inputs__ = OrderedDict()
+        num_inputs = call_fn.__code__.co_argcount
+        input_names = call_fn.__code__.co_varnames[1:num_inputs]
+        in_types = call_fn.__annotations__
+        assert set(input_names) == set(in_types.keys()), "Missing type annotations on inputs"
```
Comment (cdonovick) [inline peak/peak.py]: Should throw `TypeError` not assert.

**Comment by rdaly525** [comment]
@cdonovick, just pushed the changes

**Pull request merged**